src/services/safe_border.py

```diff
@@ -0,0 +1,316 @@
+import math
+from typing import Any, Optional, Sequence
+
+from eth_typing import HexStr
+
+from src.constants import EPOCHS_PER_SLASHINGS_VECTOR, MIN_VALIDATOR_WITHDRAWABILITY_DELAY
+from src.metrics.prometheus.duration_meter import duration_meter
+from src.modules.submodules.consensus import ChainConfig, FrameConfig
+from src.modules.accounting.typings import OracleReportLimits
+from src.utils.abi import named_tuple_to_dataclass
+from src.typings import EpochNumber, FrameNumber, ReferenceBlockStamp, SlotNumber
+from src.web3py.extensions.lido_validators import Validator
+from src.web3py.typings import Web3
+from src.utils.slot import get_blockstamp
+
+
+class WrongExitPeriod(Exception):
+    pass
+
+
+class SafeBorder:
+    chain_config: ChainConfig
+    frame_config: FrameConfig
+    blockstamp: ReferenceBlockStamp
+
+    def __init__(
+        self,
+        w3: Web3,
+        blockstamp: ReferenceBlockStamp,
+        chain_config: ChainConfig,
+        frame_config: FrameConfig,
+    ) -> None:
+        self.w3 = w3
+        self.lido_contracts = w3.lido_contracts
+
+        self.blockstamp = blockstamp
+        self.chain_config = chain_config
+        self.frame_config = frame_config
+
+        self._retrieve_constants()
+
+    @duration_meter()
+    def get_safe_border_epoch(
+        self,
+        is_bunker: bool,
+    ) -> EpochNumber:
+        if not is_bunker:
+            return self._get_default_requests_border_epoch()
+
+        negative_rebase_border_epoch = self._get_negative_rebase_border_epoch()
+        associated_slashings_border_epoch = self._get_associated_slashings_border_epoch()
+
+        return min(
+            negative_rebase_border_epoch,
+            associated_slashings_border_epoch,
+        )
+
+    def _get_default_requests_border_epoch(self) -> EpochNumber:
+        return EpochNumber(self.blockstamp.ref_epoch - self.finalization_default_shift)
+
+    def _get_negative_rebase_border_epoch(self) -> EpochNumber:
+        bunker_start_or_last_successful_report_epoch = self._get_bunker_start_or_last_successful_report_epoch()
+
+        latest_allowable_epoch = bunker_start_or_last_successful_report_epoch - self.finalization_default_shift
```

Comment:
maybe not `latest` and `earliest`, but `bunker_start_epoch` and `max_allowed_border_epoch_for_negative_rebase`?